### ISSUE TITLE
"Fallthrough is not allowed" after calling a function that never returns

### ISSUE THREAD

**Issue by metanomial** [denoland/deno_lint#782]
"Fallthrough is not allowed" is flagged after making a call to a function with the return type `never`.

```ts
switch (mode) {
  default:
    throwError("Invalid mode"); // Flagged for fallthrough
  case 1:
    handle1();
    break;
  case 2:
    handle2();
}

function throwError(message: string): never {
  throw new Error(message + debug ? " (debugging)" : "");
}
```

### Version

deno 1.12.1 (release, x86_64-unknown-linux-gnu)
v8 9.2.230.14
typescript 4.3.5

**Comment by kitsonk** [comment]
The challenge is that currently `deno_lint` cannot utilise types when linting, so it doesn't know that it is a `never` function.

**Comment by magurotuna** [comment]
Yes, as @kitsonk said, this case is pretty hard to be handled correctly by the linter due to the inability to know the return type of the function. I'd add `// deno-lint-ignore no-fallthrough` for the moment.

**Comment by jsejcksn** [comment]
I closed a duplicate issue that I opened, but wanted to quote [a comment](https://github.com/denoland/deno_lint/issues/804#issuecomment-901540809) from @magurotuna for visibility in this discussion:

> To deal with all the possible situations, type analysis is required, but unfortunately we don't have that. Maybe we should try to find a solution for simple cases like the example you described, since using functions that return `never` type in `case` statements are very commonly used technique in TypeScript.
